extension/opampextension/opamp_agent.go

```diff
@@ -0,0 +1,244 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package opampextension // import "github.com/open-telemetry/opentelemetry-collector-contrib/extension/opampextension"
+
+import (
+	"context"
+	"crypto/tls"
+	"net/http"
+	"os"
+	"runtime"
+
+	"github.com/oklog/ulid/v2"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	semconv "go.opentelemetry.io/collector/semconv/v1.18.0"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opamp-go/client"
+	"github.com/open-telemetry/opamp-go/client/types"
+	"github.com/open-telemetry/opamp-go/protobufs"
+)
+
+// TODO: Replace with https://github.com/open-telemetry/opentelemetry-collector/issues/6596
+const localConfig = `
+exporters:
+  otlp:
+    endpoint: localhost:1111
+receivers:
+  otlp:
+    protocols:
+      grpc: {}
+      http: {}
+service:
+  pipelines:
+    traces:
+      receivers: [otlp]
+      processors: []
+      exporters: [otlp]
+`
+
+type opampAgent struct {
+	cfg    *Config
+	logger *zap.Logger
+
+	agentType    string
+	agentVersion string
+
+	instanceId ulid.ULID
+
+	effectiveConfig string
+
+	agentDescription *protobufs.AgentDescription
+
+	opampClient client.OpAMPClient
+
+	remoteConfigStatus *protobufs.RemoteConfigStatus
+}
+
+func (o *opampAgent) Start(_ context.Context, _ component.Host) error {
+	o.opampClient = client.NewWebSocket(o.logger.Sugar())
```

Comment:
Should we add a TODO to support "http" transport in the future? (perhaps by looking at the `o.cfg.Endpoint` URL scheme).

Reply:
Do we intend for this extension to eventually support HTTP? I don't think we should have a TODO if it's out-of-scope.

Reply:
I think it needs to support http eventually. http transport may be my preference in some situations if I am running a Collector at massive scale and want to point them to my OpAMP backend and don't want the backend to have millions of websocket active connections.

Reply:
We also discussed this on today's Agent Management WG call. I've added a TODO, as its now clear to me that we do want to support HTTP. I've also attempted to bring the configuration structure inline with that of the supervisor, with one exception, the separation of server transport configuration (i.e. ws vs http).

Reply:
The group felt there was enough difference in settings to justify the separation, making it clear which settings will apply to a given transport. There was a suggestion to apply this separation to the Supervisor config.

Reply:
I wonder what the differences are? WebSocket is an http connection, so I would expect virtually everything applicable to http to be also applicable to WebSocket connection.

Reply:
It's also my expectation that any settings that would apply to an HTTP connection could also apply to a WebSocket connection, but the websocket library's [Dial method](https://pkg.go.dev/github.com/gorilla/websocket#Dialer.Dial) doesn't expose many options. I had assumed that the OpAMP client would allow supplying an HTTP client (which would allow us to use `HTTPClientSettings`) or would support more options, but it looks like both transports are only configurable through `types.StartSettings`.

The aim of having separate config sections for `ws` and `http` would be to make it clear to the user which options are allowed by the library used for each transport. If we plan to support supplying an HTTP client to the OpAMP client when using HTTP, then I think separate config sections will make sense. Otherwise we should just put them back into a single section.

Reply:
OK, let's keep as is for now and we can decide it later.